### PR TITLE
adds appveyor config

### PR DESCRIPTION
**Pull request by mathiask88**
Closes https://github.com/prebuild/prebuild/issues/207

@ralphtheninja

### CODE REVIEW

**Comment by ralphtheninja** [comment]
Hmm interesting. We actually got a build error there for node 4 and 32 bit.

**Comment by mathiask88** [comment]
Maybe a singularity, lets wait for the merge build.

**Comment by ralphtheninja** [comment]
Looks like it's actually failing on that particular version. Which is _good_ since we can hopefully reproduce and fix it. (I don't like temporary hickup problems)

**Comment by vweevers** [comment]
I think the problem is in the environment variables. Somewhere, all variables with a value of `undefined` are converted to the *string* `'undefined'` and then npm tries to `require('undefined')` as an [onload-script](https://docs.npmjs.com/misc/config#onload-script).

**Comment by vweevers** [comment]
Able to reproduce on Node.js 4.5.0 32 bits.

**Comment by mathiask88** [comment]
@vweevers Yes, you are right. Let's open an issue for that. But this PR should be fine I think?

**Comment by vweevers** [comment]
> Let's open an issue for that. But this PR should be fine I think?

I can only reproduce with latest npm. So can you remove `npm -g install npm` from `appveyor.yml`? Let's see what happens. I think that's safer anyway, to rely on the npm version bundled with node.

Note we may need latest npm on node 5 to fix https://github.com/nodejs/node-gyp/issues/972. See [leveldown/apppveyor.yml](https://github.com/Level/leveldown/blob/beaf17a2333a8d83d8b5df061bd748277b971ab9/appveyor.yml#L20) as an example.

**Comment by ralphtheninja** [comment]
Hmm, ok. So we're not doing anything wrong? Where is the issue?

**Comment by emilbayes** [comment]
I would also remove node.js version 5 and 7, since they're both EOL 😄 No need to be wasteful with free resources

**Comment by ralphtheninja** [comment]
@emilbayes No reason. Make a PR? :)

File: appveyor.yml [hunk 1]
```diff
@@ -29,9 +27,6 @@ platform:
 install:
   # Get the latest version of $env:nodejs_version
   - ps: Update-NodeJsInstallation (Get-NodeJsLatestBuild $env:nodejs_version) $env:platform
-  # Update to latest npm
-  - npm -g install npm 2>&1>nul
-  - SET PATH=%APPDATA%\npm;%PATH%
```
Comment (ralphtheninja) [inline appveyor.yml]: 👍

Comment (vweevers) [reply appveyor.yml]: 🤞

**Comment by ralphtheninja** [comment]
Hmm, same probem 😠

**Comment by vweevers** [comment]
There must be some random factor. Locally, I failed to reproduce with a bundled npm version, but let me see what happens if I run the test many many times.

**Comment by vweevers** [comment]
> Locally, I failed to reproduce with a bundled npm version, but let me see what happens if I run the test many many times.

No luck. And I cannot make sense of it. As a hail mary, we could change this:

https://github.com/prebuild/prebuild/blob/d820ab5daf576dea8579c654100329295d7576e1/test/integration-test.js#L13

To:

```js
exec('npm run prebuild', { cwd: cwd, env: {} }, function (error, stdout, stderr) {
```

So that the child process doesn't inherit the env.

**Comment by ralphtheninja** [comment]
> As a hail mary

😆

**Comment by vweevers** [comment]
Pff now everything is failing:

```
Error: Failed to replace env in config: ${APPDATA}
    at C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:417:13
    at String.replace (<anonymous>)
    at envReplace (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:413:12)
    at parseField (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:391:7)
    at C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:334:17
    at Array.forEach (<anonymous>)
    at Conf.add (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:333:23)
    at ConfigChain.addString (C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\config-chain\index.js:244:8)
    at Conf.<anonymous> (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:321:10)
    at C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\graceful-fs\graceful-fs.js:78:16
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:528:3)
TypeError: Cannot read property 'get' of undefined
    at errorHandler (C:\Program Files (x86)\nodejs\node_modules\npm\lib\utils\error-handler.js:205:18)
    at C:\Program Files (x86)\nodejs\node_modules\npm\bin\npm-cli.js:83:20
    at cb (C:\Program Files (x86)\nodejs\node_modules\npm\lib\npm.js:224:22)
    at C:\Program Files (x86)\nodejs\node_modules\npm\lib\npm.js:262:24
    at C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:81:7
    at Array.forEach (<anonymous>)
    at Conf.<anonymous> (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:80:13)
    at Conf.f (C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\once\once.js:25:25)
    at Conf.emit (events.js:160:13)
    at Conf.add (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:337:10)
    at ConfigChain.addString (C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\config-chain\index.js:244:8)
    at Conf.<anonymous> (C:\Program Files (x86)\nodejs\node_modules\npm\lib\config\core.js:321:10)
    at C:\Program Files (x86)\nodejs\node_modules\npm\node_modules\graceful-fs\graceful-fs.js:78:16
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:528:3)
```

**Comment by vweevers** [comment]
I'm guess Appveyor's `npmrc` has a value with `${APPDATA}` in it, expecting `APPDATA` to be defined in the environment. Sigh.

**Comment by vweevers** [comment]
@mathiask88 your initial instinct was right. I'm sorry for leading you down this dark path 😉. I suggest to:

1. Revert bef47cd4eaf8cc280c7da072b14bcf5024f6991d
2. Remove node 4 from `appveyor.yml`
3. Open an issue for it (needs a deep-dive)
4. Merge this PR.

**Comment by ralphtheninja** [comment]
This error really pisses me off because it makes _no_ sense. Why would a 32 bit version of the same node version cause errors related to environment variables?

**Comment by vweevers** [comment]
> Why would a 32 bit version of the same node version cause errors related to environment variables?

It's Windows. I did find https://github.com/nodejs/node/issues/14593, might be related.

**Comment by mathiask88** [comment]
@ralphtheninja Something similar here https://github.com/mapbox/node-pre-gyp/issues/209 ? I just looked over it quickly.

**Comment by vweevers** [comment]
You know what else fixes it for me (with latest npm, because that's the only way I can reproduce)? Merely adding the line:

```js
console.log(process.env.npm_config_onload_script)
```

Which btw, is an empty string. Just bonkers. Some quantum mechanics thing where reading the data changes the data.

**Comment by mathiask88** [comment]
I think this is nothing we will solve today. What about using [this](https://github.com/prebuild/prebuild/blob/52baa195a1e6f11613b7b1352f63affebceec268/test/integration-test.js#L10-L13) dead test code to skip this weird error?
```javascript
  if ((process.versions.modules === '46') && (process.arch === 'ia32')) {
    console.log('Skipping test on node 4.x-ia32!')
    return t.end()
  }
```

Another question is, should we run the `nspcheck` also on windows or is Linux/OSX sufficient?

**Comment by vweevers** [comment]
> Another question is, should we run the nspcheck also on windows or is Linux/OSX sufficient?

Does `nsp` do Windows-specific checks?

**Comment by ralphtheninja** [comment]
Just some info. Tried building on a separate branch on my fork. I've disabled all versions _but_ node 4 32 bit and also all tests _but_ the integration test. If I try to bypass `npm` by doing `exect(`node ../../bin.js...)` I get the same error, so it seems it not related to `npm` fwiw.

https://ci.appveyor.com/project/ralphtheninja/prebuild/build/3

**Comment by mathiask88** [comment]
> Does nsp do Windows-specific checks?

I don't know but I don't think so. At least the code never uses `process.platform`.

@ralphtheninja looks like the [prebuild variable](https://github.com/prebuild/prebuild/blob/52baa195a1e6f11613b7b1352f63affebceec268/test/integration-test.js#L16) closure does not work, because `existsSync(..)` fails with `ENOENT: no such file or directory, open 'C:\projects\prebuild\test\native-module\undefined'` and that is `cwd + undefined variable`

**Comment by vweevers** [comment]
No harm in keeping `nsp`, I think.

**Comment by ralphtheninja** [comment]
Sorry, my bad. It's not the same error. I get an error from `node-gyp` if I bypass `npm`:

```
prebuild ERR! stack Error: ENOENT: no such file or directory, open 'C:\projects\prebuild\test\native-module\undefined'
prebuild ERR! stack     at Error (native)
prebuild ERR! stack     at Object.fs.openSync (fs.js:549:18)
prebuild ERR! stack     at Object.fs.readFileSync (fs.js:397:15)
prebuild ERR! stack     at readCAFile (C:\projects\prebuild\node_modules\node-gyp\lib\install.js:466:15)
prebuild ERR! stack     at download (C:\projects\prebuild\node_modules\node-gyp\lib\install.js:438:22)
prebuild ERR! stack     at C:\projects\prebuild\node_modules\node-gyp\lib\install.js:185:19
prebuild ERR! stack     at C:\projects\prebuild\node_modules\mkdirp\index.js:30:20
prebuild ERR! stack     at FSReqWrap.oncomplete (fs.js:82:15)
prebuild ERR! not ok 
prebuild ERR! build Error: ENOENT: no such file or directory, open 'C:\projects\prebuild\test\native-module\undefined'
prebuild ERR! build     at Error (native)
prebuild ERR! build     at Object.fs.openSync (fs.js:549:18)
prebuild ERR! build     at Object.fs.readFileSync (fs.js:397:15)
prebuild ERR! build     at readCAFile (C:\projects\prebuild\node_modules\node-gyp\lib\install.js:466:15)
prebuild ERR! build     at download (C:\projects\prebuild\node_modules\node-gyp\lib\install.js:438:22)
prebuild ERR! build     at C:\projects\prebuild\node_modules\node-gyp\lib\install.js:185:19
prebuild ERR! build     at C:\projects\prebuild\node_modules\mkdirp\index.js:30:20
prebuild ERR! build     at FSReqWrap.oncomplete (fs.js:82:15)
] killed: false, code: 2, signal: null, cmd: 'node ../../bin.js --target 1.4.11 --runtime electron' }
```

**Comment by mathiask88** [comment]
@ralphtheninja It is the same [error](https://ci.appveyor.com/project/mathiask88/prebuild/build/12/job/e0kxt9lanxx7ixa9#L346) but we have two more errors before from npm

**Comment by vweevers** [comment]
@ralphtheninja @mathiask88 Yea, I think it's the same underlying bug. I've been trying to replicate it sans npm too, using https://github.com/nodejs/node/issues/14593#issuecomment-319602213 as inspiration, but no luck yet.

**Comment by ralphtheninja** [comment]
I tried disabling the code reading `cafile` https://github.com/ralphtheninja/node-gyp/commit/cd2251a8147594a91bb1779343da6ca959dc78b2 and then I can build

**Comment by ralphtheninja** [comment]
Would it be terrible if we skipped running the tests on 32 bit and only on 64 bit?

**Comment by vweevers** [comment]
> Would it be terrible if we skipped running the tests on 32 bit and only on 64 bit?

@mathiask88 suggested to skip only the failing test on 32 bit: https://github.com/prebuild/prebuild/pull/208#issuecomment-361100498, I think that's a good temporary solution.

**Comment by vweevers** [comment]
This is where I got. Minimal repro:

```js
// test.js
var exec = require('child_process').exec

if (process.argv[2] === 'block') {
  // Any blocking operation triggers the bug
  try { require('fs').readFileSync('foobar') } catch (e) {}
}

exec('node -p "process.env.npm_config_onload_script === \'undefined\' ? \'BUG\' : \'OK\'"', function (err, stdout) {
  if (err) throw err
  console.log(stdout)
})
```

````json
"scripts": {
    "test": "node test.js"
}
```` 

Output:

```
> npm test
OK
> npm test -- block
BUG
```

**Comment by ralphtheninja** [comment]
@mathiask88 Yes, lets use your hack to not run on node 4 32 bit. To your other question, I think it's ok to run nsp check on the other platforms for now. Lets just get these tests on windows through first and we can tweak it later.

**Comment by ralphtheninja** [comment]
@mathiask88 Also, if you clean up/rebase this branch onto latest master, you'll get an additional test that will have the same problem. (The `node-webkit` thingie).

**Comment by vweevers** [comment]
It's a node bug. Reproduced without npm:

```js
var exec = require('child_process').exec

process.env.test = ''

if (process.argv[2] === 'trigger') {
  try { require('fs').readFileSync('foobar') } catch (e) {}
}

exec('node -p "process.env.test === \'undefined\' ? \'BUG\' : \'OK\'"', function (err, stdout) {
  if (err) throw err
  console.log(stdout)
})
```

Output:

```
> node test.js
OK
> node test.js trigger
BUG
```

**Comment by mathiask88** [comment]
@vweevers I can repro, but still weird. If you add a `console.log('foo')` before `exec(..)` it works :see_no_evil:

**Comment by vweevers** [comment]
Left a comment on the Node.js thread: https://github.com/nodejs/node/issues/14593#issuecomment-361111588.

**Pull request merged**